### PR TITLE
Ensure server LogLevel and LogFormat have defaults (#2607)

### LINKED ISSUE #2607: Empty YAML server stanza causes debug-level logging
If `agent` is configured with an empty `server` stanza in YAML:

```yml
server:
metrics:
...

```

Then `log_level` and `log_format` will lose their defaults and be assigned empty strings, as shown by the `-/config` endpoint:

```yml
server:
  log_level: ""
  log_format: ""
  grpc_tls_config: {
...
```

The [default values are supposed to be](https://grafana.com/docs/agent/latest/configuration/server-config/) `info` and `logfmt`.

For the `log_level`, unfortunately an unknown level is mapped to "all" by the logging library; this generates *significant amounts* of unwanted log output. 
https://github.com/cortexproject/cortex/blob/v1.11.0/pkg/util/log/log.go#L124

Ironically, I triggered this issue as part of upgrading my config from `v0.24`:
```diff
 server:
-  http_listen_address: '{{ ansible_host }}'
-  http_listen_port: 9090
metrics:
``` 

Adding any other entry to the stanza alleviates the issue: 

```yml
server:
  http_tls_config:
```

I stepped through and thought about it for a bit, and I think that in the case of just `server:`, the `config.Config.Server` is getting Golang-default-initialized, which puts empty strings in those fields, whereas in the `server: http_tls_config:` case, it is specifically `config.Config.Server.HTTP.TLSConfig` that is getting default-initialized.  And of course if no `server:` is specified, there is nothing overwritten.

I think the best course of action is to specifically check for default-initialization of those fields, since Golang want `""` and Agent wants `info / logfmt`.   It is unlikely an operator would intentionally make blank and it's invalid anyway.

### PR DESCRIPTION
**Pull request by neomantra**
#### PR Description

Some YAML parsing of the Server configs will default initialize LogLevel and LogFormat to blank.  This adds a check and corrects the values to the defaults.

#### Which issue(s) this PR fixes
#2607

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Review by rfratto — COMMENTED** [review]
Hey @neomantra, thanks for the PR! Sorry how long it took for us to get back to you on this; most of the team was on holiday throughout December. 

Can you also add a CHANGELOG entry for this bug fix? This [guide](https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog) outlines what you need to do for that.

File: pkg/config/config.go [hunk 1]
```diff
@@ -113,6 +113,12 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		return err
 	}
 
+	// Ensure server logging is default-configured, some YAML parsings may unintentionally blank them.
+	if fc.Server.LogLevel.String() == "" && fc.Server.LogFormat.String() == "" {
+		fc.Server.LogLevel.Set("info")
+		fc.Server.LogFormat.Set("logfmt")
+	}
```
[the remark this answers is not in the document]
Comment (neomantra) [reply pkg/config/config.go:120]: Sorry for delay in this; thanks for the review.  I did work a bunch on it, got stalled, now pushing through.  I agree this had some code smell; your correction still left me for desiring.  I'm about to push stuff that supersedes this approach.

**Comment by neomantra** [comment]
I rebased to `main` and did a different approach...

So I was generally worried that this YAML parsing issue spread through `grafana-agent`.   After familiarizing myself with the codebase, I saw other packages used `ApplyDefaults` and `Validate` methods to handle complex initialization, which makes sense.  `Server` wasn't complex until recently; now it is with those fields.  It made sense to harmonize it, so I added `server.Config.ApplyDefaults` and invoke that in the `config.Validate`.

I added unit tests for this issue.  I also note that none of the other unit tests checked as end-to-end as `config.Load`.  I started with a string-based test case and the tests didn't fail!

The `IsZero()` aspect of your code review made me wonder why the YAML parser wasn't catching this.  I still don't totally get it after stepping a bunch.   I knew something like #2664 was needed, but didn't have the bandwidth to do it, so glad that has happened.

However, *#2664 is not enough* -- this PR is still required to alleviate the issue.   You can verify that by commenting out the body of `config.server.ApplyDefaults` and running `go test ./pkg/config`.

**Comment by neomantra** [comment]
I opted to not use the reflect approach in this because I thought it was clearer to call out the things that need complex initialization, especially now that it is scoped to the `server` package instead of `config`.  If you disagree, that's all good and I can change it.

**Comment by neomantra** [comment]
I just rebased this again, as the conflict was blocking.  Thanks!

**Comment by neomantra** [comment]
I just rebased this again, including a minor code change for the new `Load` argument introduced in this [commit](https://github.com/grafana/agent/commit/557da3a504cbf45d3243fa54d4a1b76d340fe6d2).   Also noting that I tested it (with the included unit test)and this [specific change](https://github.com/grafana/agent/commit/cad86dd606b8082a4c92414fd5b7c89845b7bef5?diff=unified#diff-67197f5fae32dd9e340a8b544610afd7add9d373e5de5cd806b1e99f032cb2c3R54) does not fix the problem (not that it was intended to).

**Comment by rfratto** [comment]
The majority of the team is out this week. @tpaschalis can you look at this when you're back next week?

**Comment by tpaschalis** [comment]
For sure! Apologies in advance for the delay @neomantra. I'm taking a look.

**Comment by tpaschalis** [comment]
Hey there 👋  I've looked through the presented solution and would like to propose an alternative if that's okay.

As you've already stated, the issue is that YAML interprets the empty block as a null one. Up until now, we've made a conscious effort to avoid assigning a special meanings to empty blocks (unlike what the OTel collector does), so I would prefer if we keep on the same pattern.

How would you feel about making the Server held the overall configuration into a pointer?

https://github.com/grafana/agent/blob/319b7a73d10f50791beef54f46bb216ac10a6547/pkg/config/config.go#L66

We can use this to then check it against nil in the Validate method and return an error `if c.Server == nil`. If the block is missing then it will happily accept its defaults.

**Comment by neomantra** [comment]
Hello and thanks again for looking at this -- no need to apologize to me ;) , this effort is for the community.

So my PR has a unit test in there to verify proper initialization and it also tests code paths that are otherwise untested.  Either way, that test should remain and any approach should pass it.

The problem arises due to YAML parsing, but it is *not* the root cause.  The root cause is that an "enum" struct member is default initialized to a logically invalid enum value. .   That is why the other `Config` objects have a `Validate / ApplyDefaults` interface.   Abstractly, this bug could occur in other ways if the resulting struct is not handled carefully.  

Your suggestion only works around this rather than addressing it directly.  I personally think it is applying it at the wrong level. 

My last change is NOT making a special exception for this YAML situation.  It is applying the same `Validate / ApplyDefaults` interface the other objects use and was missing in `server.Config`.    However, I do note that the [`config.Config.Logs` member](https://github.com/grafana/agent/blob/main/pkg/config/config.go#L70)  is a pointer type (none of the others are), but it also has `ApplyDefaults`.  

I also note that `server.Config` is in a public `pkg/server` directory and thus exposed to external developers and thus including the appropriate interface to handle its idiosyncrasies is helpful for future `grafana-agent` work and anything else leveraging that package (doubtful, but it is in the wild).

All that said, I just implemented your approach and it does pass the new-in-this-PR `TestConfig_EmptyServerConfig`.   I will clean it up and make a pull request for this approach as well, then you and your team can compare the approaches side-by-side and see which fits your project better.

**Comment by neomantra** [comment]
I have rebased this PR to `main` so that one can compare PRs #2608 and #3030 side-by-side from the same base.

**Comment by tpaschalis** [comment]
Thanks for the effort, its good to compare them side-by-side. That new approach works fine and I'm personally more fond of it as a simpler one.

> Either way, that test should remain and any approach should pass it.
> The root cause is that an "enum" struct member is default initialized to a logically invalid enum value

My original recommendation was to actually _break_ the test by making the empty server block invalid.
The reasoning the user either does _not_ provide any server block (so naturally, defaults are applied), or they _do_ provide it and can set their options. Providing a null block should not be valid behavior from a user perspective, no matter whether the underlying library handled the empty `log_level` correctly or not.

**Comment by neomantra** [comment]
OK, in that case, let's update the test and "fix" it!

**Comment by tpaschalis** [comment]
I'm closing this PR in favor of the approach in #3030

**Pull request closed without merging**